R/sugar.R

```diff
@@ -66,3 +66,157 @@ opt = function(.key, ...) {
+#'  * `value` - Optimal outcome
+#'  * `instance` - [OptimInstanceSingleCrit]
+#'
+#' @export
+#' @examples
+#' library(paradox)
+#'
+#' # function and search space bounds
+#' fun = function(xs) {
+#'   - (xs[[1]] - 2)^2 - (xs[[2]] + 3)^2 + 10
+#' }
+#'
+#' bb_optimize(
+#'   method = "random_search",
+#'   fun = fun,
+#'   term_evals = 10,
+#'   lower = c(-10, -5),
+#'   upper = c(10, 5))
+#'
+#' # function and search space
+#' fun = function(xs) {
+#'   - (xs[[1]] - 2)^2 - (xs[[2]] + 3)^2 + 10
+#' }
+#'
+#' search_space = ps(
+#'  x1 = p_dbl(-10, 10),
+#'  x2 = p_dbl(-5, 5)
+#' )
+#'
+#' bb_optimize(
+#'   method = "random_search",
+#'   fun = fun,
+#'   term_evals = 10,
+#'   search_space = search_space)
+#'
+#' # objective
+#' fun = function(xs) {
+#'   c(z = - (xs[[1]] - 2)^2 - (xs[[2]] + 3)^2 + 10)
+#' }
+#'
+#' search_space = ps(
+#'  x1 = p_dbl(-10, 10),
+#'  x2 = p_dbl(-5, 5)
+#' )
+#'
+#' codomain = ps(
+#'  z = p_dbl(tags = "minimize")
+#' )
+#'
+#' objective = ObjectiveRFun$new(fun, search_space, codomain)
+#'
+#' bb_optimize(
+#'   method = "random_search",
+#'   objective = objective,
+#'   term_evals = 10)
+#' 
+bb_optimize = function(method, fun = NULL, objective = NULL, term_evals = NULL, term_time = NULL, search_space = NULL,
+  lower = NULL, upper = NULL, codomain = NULL, ...) {
+
+  assert_choice(method, mlr_optimizers$keys())
+  optimizer = opt(method, ...)
+  assert_function(fun, null.ok = TRUE)
+  assert_r6(objective, "Objective", null.ok = TRUE)
+  terminator = terminator_selection(term_evals, term_time)
+  if (!is.null(search_space)) assert_param_set(search_space)
+  assert_numeric(lower, null.ok = TRUE)
+  assert_numeric(upper, null.ok = TRUE)
+
+  if (!xor(is.null(fun), is.null(objective))) {
+    stop("Either `fun` or `objective` must be provided.")
+  }
+
+  if (xor(is.null(lower), is.null(upper))) {
+    stop("`lower` and `upper` must be provided.")
+  }
+
+  if (!is.null(fun)) {
+    if(is.null(search_space)) {
+      ids = if (is.null(names(lower))) paste0("x", seq(lower)) else names(lower)
+
+      search_space = ParamSet$new(pmap(list(ids, lower, upper), function(s, lb, ub) {
+        ParamDbl$new(id = s, lower = lb, upper = ub)
+      }))
+    }
+    if (is.null(codomain)) codomain = ps(y = p_dbl(tags = "minimize"))
+    objective = ObjectiveRFun$new(fun, search_space, codomain, check_values = FALSE)
+  }
+
+  instance = OptimInstanceSingleCrit$new(objective, terminator = terminator, check_values = FALSE)
+  optimizer$optimize(instance)
+  par = if (instance$search_space$all_numeric) {
+    set_names(as.numeric(instance$result_x_search_space), instance$search_space$ids()) 
+  } else {
+    instance$result_x_search_space
+  }
+
+  list(par =  par, value = instance$result_y, instance = instance)
+}
+
+terminator_selection = function(term_evals, term_time) {
+  assert_int(term_evals, null.ok = TRUE)
+  assert_int(term_time, null.ok = TRUE)
+
+  if (is.null(term_evals) && is.null(term_time)) {
+    trm("none")
```

Comment:
Does this default make sense? Would this mean that we probably would never stop?

Reply:
Not anymore if we use random search as the default optimization method. Maybe we should use `max_evals = 100`?

Reply:
Rather `1000`. I would use this mainly for relatively cheap function optim where this is usually a reasonable default.

Reply:
Changed to 1000